reuse.go

```diff
@@ -0,0 +1,140 @@
+package libp2pquic
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/vishvananda/netlink"
+)
+
+type reuseConn struct {
+	net.PacketConn
+	refCount int32 // to be used as an atomic
+}
+
+func newReuseConn(conn net.PacketConn) *reuseConn {
+	return &reuseConn{PacketConn: conn}
+}
+
+func (c *reuseConn) IncreaseCount() { atomic.AddInt32(&c.refCount, 1) }
+func (c *reuseConn) DecreaseCount() { atomic.AddInt32(&c.refCount, -1) }
+func (c *reuseConn) GetCount() int  { return int(atomic.LoadInt32(&c.refCount)) }
+
+type reuse struct {
+	mutex sync.Mutex
+
+	unicast map[string] /* IP.String() */ map[int] /* port */ *reuseConn
+	// global contains connections that are listening on 0.0.0.0 / ::
+	global map[int]*reuseConn
+}
+
+func newReuse() *reuse {
+	return &reuse{
+		unicast: make(map[string]map[int]*reuseConn),
+		global:  make(map[int]*reuseConn),
+	}
+}
+
+func (r *reuse) getSourceIPs(network string, raddr *net.UDPAddr) ([]net.IP, error) {
+	// Determine the source address that the kernel would use for this IP address.
+	// Note: This only works on Linux.
+	// On other OSes, this will return a netlink.ErrNotImplemetned.
+	routes, err := (&netlink.Handle{}).RouteGet(raddr.IP)
+	if err != nil {
+		return nil, err
+	}
+
+	ips := make([]net.IP, 0, len(routes))
+	for _, route := range routes {
+		ips = append(ips, route.Src)
+	}
+	return ips, nil
+}
+
+func (r *reuse) Dial(network string, raddr *net.UDPAddr) (*reuseConn, error) {
+	conn, err := r.dial(network, raddr)
+	if err != nil {
+		return nil, err
+	}
+	conn.IncreaseCount()
+	return conn, nil
+}
+
+func (r *reuse) dial(network string, raddr *net.UDPAddr) (*reuseConn, error) {
+	ips, err := r.getSourceIPs(network, raddr)
+	if err != nil && err != netlink.ErrNotImplemented {
+		return nil, err
+	}
+
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	for _, ip := range ips {
+		// We already have at least one suitable connection...
+		if conns, ok := r.unicast[ip.String()]; ok {
+			// ... we don't care which port we're dialing from. Just use the first.
+			for _, c := range conns {
+				return c, nil
+			}
+		}
+	}
+
+	// Use a connection listening on 0.0.0.0 (or ::).
+	// Again, we don't care about the port number.
+	for _, conn := range r.global {
+		return conn, nil
+	}
+
+	// We don't have a connection that we can use for dialing.
+	// Dial a new connection from a random port.
+	var addr *net.UDPAddr
+	switch network {
+	case "udp4":
+		addr, err = net.ResolveUDPAddr("udp4", "0.0.0.0:0")
+	case "udp6":
+		addr, err = net.ResolveUDPAddr("udp6", "[::]:0")
+	}
```

Comment:
```go
net.UDPAddr{
  IP: net.IPv6zero // or net.IPv4zero,
  Port: 0,
  Network: "udpN",
}
```
?

Reply:
There's no `Network` in the `net.UDPAddr`, but I guess just `&net.UDPAddr{IP: net.IPv6zero, Port: 0}` should be fine.